eessi/testsuite/tests/apps/espresso/src/madelung.py

```diff
@@ -0,0 +1,152 @@
+        for l in range(lattice_size[2]):
+            _ = system.part.add(pos=np.multiply([j, k, l], base_vector),
+                                q=(-1.)**(j + k + l), fix=3 * [True])
+
+# setup P3M algorithm
+algorithm = espressomd.electrostatics.P3M
+if args.gpu:
+    algorithm = espressomd.electrostatics.P3MGPU
+solver = algorithm(prefactor=1., accuracy=1e-6)
+if (espressomd.version.major(), espressomd.version.minor()) == (4, 2):
+    system.actors.add(solver)
+else:
+    system.electrostatics.solver = solver
+
+
+print("Algorithm executed. \n")
+
+atol_energy = atol_pressure = 1e-12
+atol_forces = 1e-5
+atol_abs_forces = 2e-6
+
+rtol_energy = 5e-6
+rtol_pressure = 2e-5
+rtol_forces = 0.
+rtol_abs_forces = 0.
+# run checks
+forces = np.copy(system.part.all().f)
+energy, p_scalar, p_tensor = get_normalized_values_per_ion(system)
+ref_energy, ref_pressure = get_reference_values_per_ion(base_vector)
+np.testing.assert_allclose(energy, ref_energy, atol=atol_energy, rtol=rtol_energy)
+np.testing.assert_allclose(p_scalar, np.trace(ref_pressure) / 3.,
+                           atol=atol_pressure, rtol=rtol_pressure)
+np.testing.assert_allclose(p_tensor, ref_pressure, atol=atol_pressure, rtol=rtol_pressure)
+np.testing.assert_allclose(forces, 0., atol=atol_forces, rtol=rtol_forces)
+np.testing.assert_allclose(np.median(np.abs(forces)), 0., atol=atol_abs_forces, rtol=rtol_abs_forces)
+
+
+print("Executing sanity checks...\n")
+if (np.all([np.allclose(energy, ref_energy, atol=atol_energy, rtol=rtol_energy),
+        np.allclose(p_scalar, np.trace(ref_pressure) / 3.,
+                           atol=atol_pressure, rtol=rtol_pressure),
+        np.allclose(p_tensor, ref_pressure, atol=atol_pressure, rtol=rtol_pressure),
+        np.allclose(forces, 0., atol=atol_forces, rtol=rtol_forces),
+            np.allclose(np.median(np.abs(forces)), 0., atol=atol_abs_forces, rtol=rtol_abs_forces)])):
+    print("Final convergence met with tolerances: \n\
+            energy: ", atol_energy, "\n\
+            p_scalar: ", atol_pressure, "\n\
+            p_tensor: ", atol_pressure, "\n\
+            forces: ", atol_forces, "\n\
+            abs_forces: ", atol_abs_forces, "\n")
+else:
+    print("At least one parameter did not meet the tolerance, see the log above.\n")
+
+print("Sampling runtime...\n")
+# sample runtime
+n_steps = 10
+timings = []
+for _ in range(10):
+    tick = time.time()
+    system.integrator.run(n_steps)
+    tock = time.time()
+    timings.append((tock - tick) / n_steps)
+
+# write results to file
+header = '"mode","cores","mpi.x","mpi.y","mpi.z","particles","mean","std"\n'
+report = f'"{"weak scaling" if args.weak_scaling else "strong scaling"}",{n_cores},{node_grid[0]},{node_grid[1]},{node_grid[2]},{len(system.part)},{np.mean(timings):.3e},{np.std(timings, ddof=1):.3e}\n'
+print(report)
+if pathlib.Path(args.output).is_file():
+    header = ""
+with open(args.output, "a") as f:
+    f.write(header + report)
```

Comment:
This write operation is superfluous if the ReFrame runner captures stdout.

Reply:
Yes, I have to remove this and is a part of my TODO. :)